patches/tModLoader/Terraria/ModLoader/Mod.Internals.cs

```diff
@@ -278,91 +284,24 @@ partial class Mod
-				if (type.IsAbstract || type.GetConstructor(new Type[0]) == null)//don't autoload things with no default constructor
-				{
-					continue;
-				}
-				if (type.IsSubclassOf(typeof(ModItem))) {
-					AutoloadItem(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalItem))) {
-					AutoloadGlobalItem(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModPrefix))) {
-					AutoloadPrefix(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModDust))) {
-					AutoloadDust(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModTile))) {
-					AutoloadTile(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalTile))) {
-					AutoloadGlobalTile(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModTileEntity))) {
-					AutoloadTileEntity(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModWall))) {
-					AutoloadWall(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalWall))) {
-					AutoloadGlobalWall(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModProjectile))) {
-					AutoloadProjectile(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalProjectile))) {
-					AutoloadGlobalProjectile(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModNPC))) {
-					AutoloadNPC(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalNPC))) {
-					AutoloadGlobalNPC(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModPlayer))) {
-					AutoloadPlayer(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModBuff))) {
-					AutoloadBuff(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalBuff))) {
-					AutoloadGlobalBuff(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModMountData))) {
-					AutoloadMountData(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModGore))) {
+				if (type == modType){continue;}
+				if (type.IsAbstract){continue;}
+				if (type.GetConstructor(new Type[0]) == null){continue;}//don't autoload things with no default constructor
+
+				if (type.IsSubclassOf(typeof(ModGore))) {
 					modGores.Add(type);
 				}
 				else if (type.IsSubclassOf(typeof(ModSound))) {
 					modSounds.Add(type);
 				}
-				else if (type.IsSubclassOf(typeof(ModWorld))) {
-					AutoloadModWorld(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModUgBgStyle))) {
-					AutoloadUgBgStyle(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModSurfaceBgStyle))) {
-					AutoloadSurfaceBgStyle(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalBgStyle))) {
-					AutoloadGlobalBgStyle(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModWaterStyle))) {
-					AutoloadWaterStyle(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModWaterfallStyle))) {
-					AutoloadWaterfallStyle(type);
-				}
-				else if (type.IsSubclassOf(typeof(GlobalRecipe))) {
-					AutoloadGlobalRecipe(type);
-				}
-				else if (type.IsSubclassOf(typeof(ModCommand))) {
-					AutoloadCommand(type);
+				else if (typeof(ILoadable).IsAssignableFrom(type)) {
```

Comment:
heh, the rest of this chain uses `IsSubclassOf`

Reply:
`IsSubclassOf` doesn't work the same for interfaces. IIRC, when used on an interface it will determine if one interface is derived from another. When you need to check if a class implements an interface, you need to use `IsAssignableFrom`.

Reply:
Yup I remember that too